Add a default.nix and .envrc (use with direnv)

vdemeester opened the pull request:
This add two cool things if you use `nix` and `direnv`

- a `default.nix` to setup the environment needed for the project. Use `nix-shell` in the repository to make sure you have all the requirements :angel: 
- a `.envrc` that tells `direnv` to use nix to setup the environment and thus get the requirement if needed.

```bash
$ mvn
The program ‘mvn’ is currently not installed. You can install it by typing:
  nix-env -iA nixos.maven
$ cd junit5-docker
direnv: loading .envrc                                                                                                                                                                                    
these derivations will be built:
  /nix/store/0n78kk6p3dq1mllx2sbn182za1r5788z-junit5-docker.drv
these paths will be fetched (16.18 MiB download, 21.34 MiB unpacked):
  /nix/store/2xk6hbsjmayfv1vppwasmmr68a5zkl4q-apache-maven-3.3.9
  /nix/store/g4f1bms36hgg5abfd8xc4bj6sfzsy61d-bash-4.3-p48-info
  /nix/store/i76bwv541a7zj576hkdia62169r2nl0z-bash-4.3-p48-doc
fetching path ‘/nix/store/i76bwv541a7zj576hkdia62169r2nl0z-bash-4.3-p48-doc’...
fetching path ‘/nix/store/g4f1bms36hgg5abfd8xc4bj6sfzsy61d-bash-4.3-p48-info’...
fetching path ‘/nix/store/2xk6hbsjmayfv1vppwasmmr68a5zkl4q-apache-maven-3.3.9’...

*** Downloading ‘https://cache.nixos.org/nar/0wd12c00ga4rx4chnr4gq2jrk2myws2pxfpy8hiyrcj7hx5ffw3s.nar.xz’ (signed by ‘cache.nixos.org-1’) to ‘/nix/store/g4f1bms36hgg5abfd8xc4bj6sfzsy61d-bash-4.3-p48-info’...

*** Downloading ‘https://cache.nixos.org/nar/0b35cwnng6lqa7sk5i952zjxbzlwl0i4q7bmziyq39glmckfvzb2.nar.xz’ (signed by ‘cache.nixos.org-1’) to ‘/nix/store/2xk6hbsjmayfv1vppwasmmr68a5zkl4q-apache-maven-3.3.9’...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Up
*** Downloading ‘https://cache.nixos.org/nar/13hi6fl3kphqxz18k9hkas15bmrvh2n4dy3js4pp4skj7s390w6g.nar.xz’ (signed by ‘cache.nixos.org-1’) to ‘/nix/store/i76bwv541a7zj576hkdia62169r2nl0z-bash-4.3-p48-doc’...
load   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  105k  100  105k    0     0  72928      0  0:00:01  0:00:01 --:--:-- 72955

100  339k  100  339k    0     0   148k      0  0:00:02  0:00:02 --:--:--  148k

direnv: ([direnv export zsh]) is taking a while to execute. Use CTRL-C to give up.
direnv: ([/nix/store/03rb81kiwp5s4wkd60k1hwlcb7a335if-direnv-2.9.0/bin/direnv apply_dump /dev/fd/63]) is taking a while to execute. Use CTRL-C to give up.
100 8063k  100 8063k    0     0   227k      0  0:00:35  0:00:35 --:--:--  248k

building path(s) ‘/nix/store/850cv3w0jp4xvs4zhywn2wdfpc88dawy-junit5-docker’
created 47 symlinks in user environment
direnv: export +CC +CONFIG_SHELL +CXX +IN_NIX_SHELL +JAVA_HOME +NIX_BUILD_CORES +NIX_BUILD_TOP +NIX_CC +NIX_CFLAGS_COMPILE +NIX_ENFORCE_NO_NATIVE +NIX_LDFLAGS +NIX_STORE +SOURCE_DATE_EPOCH +TEMP +TEMPDIR +TMP +TMPDIR +_PATH +buildInputs +builder +env +name +nativeBuildInputs +out +propagatedBuildInputs +propagatedNativeBuildInputs +shell +stdenv +system ~PATH
$ mvn verify
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building junit5-docker 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
# […]
```

/cc @FaustXVI 

- [x] I need to add docker (and some info still)

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented on default.nix:
```diff
@@ -0,0 +1,11 @@
+stdenv.mkDerivation rec {
+    name = "junit5-docker";
+    env = buildEnv { name = name; paths = buildInputs; };
+    buildInputs = [
```
Should add `docker` here too I think :angel:

FaustXVI replied:
Definitely yes :)

FaustXVI commented on default.nix:
```diff
@@ -0,0 +1,11 @@
+with import <nixpkgs> {};
+
+# FIXME(vdemeester) should pin the channel we use (to make it even more reproductible)
```
I even would say it's mandatory !

vdemeester committed:
Add a default.nix and .envrc (use with direnv)

Using pinned version from release-16.09. Update this from time to time
to get security updates for example.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>
